### PR TITLE
 Add payment rest endpoint

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh/issues/5856

### CODE REVIEW

File: de.metas.business/src/main/java/de/metas/order/model/interceptor/C_Order.java [hunk 1]
```diff
@@ -234,10 +236,18 @@ public void linkWithPaymentByExternalOrderId(@NonNull final I_C_Order order)
...
+		// 		and since the order is not yet saved => only C_DocType_Target_ID has value and C_DocType_ID doesn't (is 0).
+		// Therefore we use this save to flush the order.
+		Services.get(IOrderDAO.class).save(order);
+
 		final I_C_Payment payment = paymentOptional.get();
-		payment.setC_Order_ID(order.getC_Order_ID());
 		order.setC_Payment_ID(payment.getC_Payment_ID());
+		payment.setC_Order_ID(order.getC_Order_ID());
 
+		DB.getSQLValue(ITrx.TRXNAME_ThreadInherited, "select C_DocType_ID from C_Order where C_Order_ID=1000004");
```
Comment (teosarca) [inline de.metas.business/src/main/java/de/metas/order/model/interceptor/C_Order.java]: dude!!!

File: de.metas.adempiere.adempiere/base/src/main/java/de/metas/location/geocoding/provider/googlemaps/GoogleMapsGeocodingProviderImpl.java [hunk 2]
```diff
@@ -103,6 +103,11 @@ public GoogleMapsGeocodingProviderImpl(final GeoApiContext context, final int ca
 		//noinspection ConfusingArgumentToVarargsMethod
 		logger.trace("Geocoding response from google: {}", results);
 
+		if (results == null)
+		{
+			return ImmutableList.of();
+		}
+
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/de/metas/location/geocoding/provider/googlemaps/GoogleMapsGeocodingProviderImpl.java]: piggybacking, but fine

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java [hunk 3]
```diff
@@ -93,7 +93,7 @@ public JsonPaymentService(final CurrencyService currencyService, final BpartnerP
 			return ResponseEntity.unprocessableEntity().body(String.format("Cannot find Bank Account for bpartner: %s, currency: %s and account: %s", jsonPaymentInfo.getBpartnerIdentifier(), jsonPaymentInfo.getCurrencyCode(), jsonPaymentInfo.getTargetIBAN()));
 		}
 
-		final I_C_Payment payment = paymentBL.newOutboundPaymentBuilder()
+		final I_C_Payment payment = paymentBL.newInboundReceiptBuilder()
```
Comment (teosarca) [inline de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java]: is this always a payment receipt, how can we make sure?

Comment (TheBestPessimist) [reply de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java]: We're not sure, but marking it as Inbound was the only way to have it read by allocation in `de.metas.allocation.modelvalidator.C_Invoice.allocateAvailablePayments(I_C_Invoice)`

Comment (TheBestPessimist) [reply de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java]: Confirmed with @metas-ts it's ok for now to only create receipts.

**Pull request merged**